### PR TITLE
add guards around some calls to memset to silence ubsan

### PR DESCRIPTION
**Pull request by RobertHenry6bev**

### CODE REVIEW

File: src/coreclr/vm/appdomain.cpp [hunk 1]
```diff
@@ -3337,7 +3337,9 @@ void AppDomain::AddUnmanagedImageToCache(LPCWSTR libraryName, NATIVE_LIBRARY_HAN
 
     size_t len = (wcslen(libraryName) + 1) * sizeof(WCHAR);
     AllocMemHolder<WCHAR> copiedName(GetLowFrequencyHeap()->AllocMem(S_SIZE_T(len)));
-    memcpy(copiedName, libraryName, len);
+    if (len && libraryName) {
```
Comment (jkotas) [inline src/coreclr/vm/appdomain.cpp:3340]: This should not be needed. We are always copying non-empty block here.

File: src/coreclr/vm/appdomain.cpp [hunk 2]
```diff
@@ -4177,7 +4179,9 @@ void DomainLocalModule::EnsureDynamicClassIndex(DWORD dwID)
         (void*)GetDomainAssembly()->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(
             S_SIZE_T(sizeof(DynamicClassInfo)) * S_SIZE_T(aDynamicEntries));
 
-    memcpy(pNewDynamicClassTable, m_pDynamicClassTable, sizeof(DynamicClassInfo) * m_aDynamicEntries);
+    if (m_aDynamicEntries && m_pDynamicClassTable) {
```
Comment (jkotas) [inline src/coreclr/vm/appdomain.cpp:4182]:
```suggestion
    if (m_aDynamicEntries != 0)
    {
```

File: src/coreclr/vm/appdomain.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/appdomain.cpp:3340]: We also don't need to check for `libraryName` as it was already validated with the call to `wcslen`.

**Review by jkotas — APPROVED** [review]
Thanks

**Comment by jkotas** [comment]
Build break:
```
D:\a\_work\1\s\src\coreclr\inc\volatile.h(453): error C2220: the following warning is treated as an error
D:\a\_work\1\s\src\coreclr\vm\appdomain.cpp(4180): note: see reference to function template instantiation 'bool Volatile<DWORD>::operator !=<int>(const TOther &) const' being compiled
```

**Pull request merged**